keystore/src/connection/platform/wasm/migrations.rs

```diff
@@ -0,0 +1,406 @@
+
+const fn db_version_number(counter: u32) -> u32 {
+    // This is how version numbers were calculated when the DB version was tied to core crypto versions
+    // 10_000_000 for a major version, 1_000 for a patch version. I.e., the number for v1.0.2 was
+    const VERSION_1_0_2: u32 = 10_000_000 + 2_000;
+    // From post v1.0.2, we will just increment whenever we need a DB migration.
+    VERSION_1_0_2 + counter
+}
+
+const DB_VERSION_0: u32 = db_version_number(0);
+
+/// Open an existing idb database with the given name and key, and migrate it if needed.
+pub(crate) async fn open_and_migrate(name: &str, key: &str) -> CryptoKeystoreResult<Database> {
+    /// Increment when adding a new migration.
+    const TARGET_VERSION: u32 = db_version_number(1);
+    let factory = Factory::new()?;
+
+    let open_existing = factory.open(name, None)?;
+    let existing_db = open_existing.await?;
+    let mut version = existing_db.version()?;
+    if version == TARGET_VERSION {
+        // Migration is not needed, just return existing db
+        Ok(existing_db)
+    } else {
+        // Migration is needed
+        existing_db.close();
+
+        while version < TARGET_VERSION {
+            version = do_migration_step(version, name, key).await?;
+        }
+
+        let open_request = factory.open(name, Some(TARGET_VERSION))?;
+        open_request.await.map_err(Into::into)
+    }
+}
+
+/// This is called from a while loop. The `from` argument represents the version the migration is performed from.
+/// The function will return the version number of the DB resulting from the migration.
+///
+/// To add a new migration, add a new match arm below the latest one.
+/// It must match on the version it migrates from, and call a function that performs the migration workload,
+///     which returns the version it migrates to, which is the same value as TARGET_VERSION in the function above at the time the migration is added.
+///
+/// However, do not use the constant but hardcode the value into the function.
+/// This way it will keep working once a new migration is added after it.
+async fn do_migration_step(from: u32, name: &str, key: &str) -> CryptoKeystoreResult<u32> {
+    match from {
+        // The version that results from the latest migration must match TARGET_VERSION
+        //      to ensure convergence of the while loop this is called from.
+        0..=DB_VERSION_0 => migrate_to_version_1(name, key).await,
+        _ => Err(CryptoKeystoreError::MigrationNotSupported(from)),
+    }
+}
+
+/// With the current feature set of stable rust macros, we're not aware how to construct an identifier
+///     for each entity inside the macro.
+/// We need it for a variable to store the conversion result.
+/// So we have to take an identifier for each entity as an argument.
+/// Can be done better once std::concat_idents is stabilized, or we can use tuple indexing when ${index()} is stabilized (https://github.com/rust-lang/rust/pull/122808).
```

Comment:
Could you please break this line into two? It's currently 169 characters long, which is a bit too much.